pkg/glusterblock/controllerserver.go

```diff
@@ -0,0 +1,316 @@
+func (cs *ControllerServer) ParseCreateVolRequest(req *csi.CreateVolumeRequest) (*ProvisionerConfig, error) {
+
+	var reqConf ProvisionerConfig
+	var gdReq api.VolCreateReq
+
+	reqConf.gdVolReq = &gdReq
+
+	// Get Volume name
+	if req != nil {
+		reqConf.gdVolReq.Name = req.Name
+	}
+	return &reqConf, nil
+}
+
+// CreateVolume creates and starts the volume
+func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) { //nolint: gocyclo
+	glog.V(2).Infof("request received %+v", req)
+
+	if err := cs.validateCreateVolumeReq(req); err != nil {
+		return nil, err
+	}
+
+	glog.V(1).Infof("creating volume with name %s", req.Name)
+
+	volSizeBytes := cs.getVolumeSize(req)
+
+	// parse the request.
+	parseResp, err := cs.ParseCreateVolRequest(req)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "failed to parse request")
+	}
+
+	volumeName := parseResp.gdVolReq.Name
+
+	blockVolInfo := bapi.BlockVolumeInfo{
+		Name: volumeName,
+		Size: uint64(volSizeBytes),
+	}
+	volumeReq := bapi.BlockVolumeCreateRequest{
+		BlockVolumeInfo: &blockVolInfo,
+	}
+
+	glog.V(2).Infof("block volume create request: %+v", volumeReq)
+	blockVolCreateResp, err := cs.client.BlockVolumeCreate("gluster-loopback", volumeReq)
+	if err != nil {
+		glog.Errorf("failed to create block volume: %s err: %v", volumeName, err)
+		errResp := cs.client.LastErrorResponse()
+		// errResp will be nil in case of `No route to host` error
+		if errResp != nil && errResp.StatusCode == http.StatusConflict {
+			return nil, status.Errorf(codes.Aborted, "block volume create already in process: %v", err)
+		}
+
+		return nil, status.Errorf(codes.Internal, "failed to create block volume: %s err: %v", volumeName, err)
+	}
+
+	glusterServer, bkpServers, err := utils.GetClusterNodes(cs.client)
+	if err != nil {
+		glog.Errorf("failed to get cluster nodes: %v", err)
+		return nil, status.Errorf(codes.Internal, "failed to get cluster nodes: %v", err)
+	}
+
+	glog.V(3).Infof("blockvolume create response : %+v", blockVolCreateResp)
+	resp := &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			VolumeId:      volumeName,
+			CapacityBytes: volSizeBytes,
+			VolumeContext: map[string]string{
+				"glustervol":        blockVolCreateResp.HostingVolume,
+				"glusterserver":     glusterServer,
+				"glusterbkpservers": strings.Join(bkpServers, ":"),
+			},
+		},
+	}
+
+	glog.V(4).Infof("CSI block volume response: %+v", resp)
+	return resp, nil
+}
+
+func (cs *ControllerServer) getVolumeSize(req *csi.CreateVolumeRequest) int64 {
+	// If capacity mentioned, pick that or use default size 1 GB
+	volSizeBytes := defaultVolumeSize
+	if capRange := req.GetCapacityRange(); capRange != nil {
+		volSizeBytes = capRange.GetRequiredBytes()
+	}
+	return volSizeBytes
+}
+
+func (cs *ControllerServer) validateCreateVolumeReq(req *csi.CreateVolumeRequest) error {
```

Comment:
i  think some of the common validation functions can be moved to a common place, this can be done as part of cleanup also

Reply:
I will do code refactor in separate patch